### PR TITLE
Add Support for View of vector of Ptr

### PR DESCRIPTION
**Pull request by wddgit**
Add support for View of vector of Ptr. This completes implementation of this new feature. There is a test of its usage in the class ViewAnalyzer in function testStdVectorPtr which is run by ViewTest_cfg.py in the package FWCore/Integration. This does change interfaces. Code in the release was fixed to compile and so that runTheMatrix will pass. But runtime problems could exist in code not run in runThe Matrix and compile time problems for code not in the release. For example, it is no longer possible to get the container product from a Ref which implies many things. For example on cannot construct a RefProd from a Ref. One change in particular that might cause problems. If one inserts into an AssociationMap constructed with the no argument constructor an exception will be thrown at runtime, There is known code like this in the release, although none is called in runTheMatrix tests.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Please test

**Comment by Dr15Jones** [comment]
+1

**Comment by slava77** [comment]
Hi David, 
Could you please add a bit longer description in the subject of the PR:
- is the feature of View of Ptr complete now? 
- are there tests or examples in this PR testing or showing that?

**Comment by Dr15Jones** [comment]
The test failure is unrelated to this pull request since it has shown up in other tests.

**Comment by slava77** [comment]
@cmsbuild please test

CMSSW_7_5_X_2015-03-26-2300 is available now and it appears to be clean from build errors
(compared to 35 errors in CMSSW_7_5_X_2015-03-26-1100)

**Comment by slava77** [comment]
a merge conflict is here now
Once that's resolved, maybe we should hold off merging other things that would interfere?

**Comment by wddgit** [comment]
I'll look for the merge conflict and try to fix it now.

Note, this will not auto merge into ROOT5 cleanly. I will fix this after the auto merge occurs with a pull request to the ROOT5 branch.

**Comment by slava77** [comment]
Hi David,

Sorry, I didn't mean to suggest to have the subject of the PR so long, just the description (the first item that shows up in a larger window when a PR is submitted or the first window below the PR name)

**Comment by slava77** [comment]
nm, you fixed it. 
Thanks

File: DataFormats/TrackReco/interface/TrackExtraBase.h [hunk 1]
```diff
@@ -51,8 +51,13 @@ class TrackExtraBase
...
     /// get a ref to i-th recHit
-    TrackingRecHitRef recHitRef(unsigned int i) const {                                                               
-        return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
+    TrackingRecHitRef recHitRef(unsigned int i) const {
+      if(m_hitCollection.productPtr()) {
+        TrackingRecHitRef::finder_type finder;
+        TrackingRecHitRef::value_type const* item = finder(*(static_cast<TrackingRecHitRef::product_type const*>(m_hitCollection.productPtr())), m_firstHit+i);
+        return TrackingRecHitRef(m_hitCollection.id(), item, m_firstHit+i);
+      }
+      return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
```
Comment (VinInn) [inline DataFormats/TrackReco/interface/TrackExtraBase.h]: do not understand the logic here:
 what is the difference between what inside the bracket and the default return?

Comment (wddgit) [inline DataFormats/TrackReco/interface/TrackExtraBase.h]: Ref has changed. The cached pointer it contains used to point at the container but now it points at the element in the container (much more similar to the way Ptr works and this allows Ptr to Ref conversions). The cached pointer in the RefProd continues to point at the container and m_hitCollection has a RefCore from a RefProd, so the pointer needs to be converted from pointing to the container to pointing at the element in this case.

**Comment by wddgit** [comment]
The merge conflict should be resolved now. It was created by unrelated code merged into 7_5_X since I submitted the PR yesterday. I did not retest things but it was pretty trivial to resolve so I will leave retesting to Jenkins.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

File: DataFormats/TrackReco/interface/TrackExtraBase.h (hunk 1 above)
Comment (VinInn) [inline DataFormats/TrackReco/interface/TrackExtraBase.h]: ok for what in the bracket (a bit convoluted, but ok)
what about the default
```
+      return TrackingRecHitRef(m_hitCollection,m_firstHit+i);
```
?

Comment (wddgit) [inline DataFormats/TrackReco/interface/TrackExtraBase.h]: The Ref contains a transient pointer as a data member. Sometimes it points at the element (in earlier versions of the code it used to point at the product), but when the Ref has just been read from an input file (and in some other circumstances) the pointer points at an EDProductGetter (a pointer which is set in a custom streamer). The EDProductGetter can get the product from the Event using the ProductID that the Ref also contains. That part of Ref has not changed.  If the function productPtr returns null, then that means the pointer in the RefCore is pointing at the EDProductGetter. And a Ref in that state can be constructed from just a RefCore and a key. In that state, RefProd and Ref contain the same pointer to the EDProductGetter. Somewhat complicated. Ask again if I did not explain it clearly enough.

**Comment by ktf** [comment]
This is a genuine mistake, I guess. Should be `==` no?

**Comment by slava77** [comment]
@cmsbuild please test
the fix to the issue in PhysicsTools/PatAlgos is in  IB by now

**Comment by wddgit** [comment]
@ktf  @wmtan What is this comment referring to?

"This is a genuine mistake, I guess. Should be == no?"

I am guessing maybe it is referring to the llvm compilation
warning?  It looks like instead of RI != RI the code should be
RI != RE and it is a bug.

But this is unrelated to this PR and in the cut parser and expression
parser code. Maybe we can ignore this warning
for this PR, although someone should follow up on it and
fix it.

Also looking through this, I think the status is that we are waiting
on the tests to run again. As far as I can see this has happened
so far

1. An compilation error in PatAlgos unrelated to this PR. 
2. It was fixed and tests rerun resulting in an llvm warning
also unrelated to this PR.
3. I do not think the llvm warning was fixed, but I am not sure?
4. Tests restarted again and still running although I am not
sure why. It looks like the previous tests passed except for
the unrelated llvm warning.

Is there anything I need to do to move this forward? Did any
tests fail that I am not noticing?

**Comment by slava77** [comment]
it looks like jenkins is still compiling this PR, after my earlier request today.

**Comment by wmtan** [comment]
@wddgit @ktf 
It certainly looks like "RI != RI" is wrong and should be "RI != RE".  I will investigate.  But this has nothing whatever to do with this PR, as @wddgit has said.

**Comment by deguio** [comment]
+1
DQM only slightly touched

**Comment by nclopezo** [comment]
Hi all,
Github is being attacked and this is causing issues with the pr tests.

**Comment by wddgit** [comment]
I believe the unit test failure was unrelated to my changes and I could not reproduce it locally. But I tried to fix it anyway and committed the fix. Can someone try the tests again? The failing unit test was writing a file called test.root and placing it in the top level directory. I'm guessing maybe some unrelated test running in parallel was stomping on the file so I gave it a more unique name and the test is now running in the tmp subdirectory. Since I do not know how to reproduce the problem I cannot test if I fixed it.

**Comment by slava77** [comment]
I was looking at the matrix logs and they looked like a frontier timeout:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/2421/runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECO+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECO+HARVEST+ALCATT.log
the step3 job started at 19:44 and never got past the first event after starting it at 20:43:38.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/8559	36f7f53
code changes are in line with expectations at visual inspection.

jenkins tests of RECO workflows pass OK and show no differences in monitored quantities (fwlite or DQM based comparisons).

**Comment by wddgit** [comment]
Thanks Slava.  Can I suggest other people go ahead and review/approve this? I was hoping for the jenkins tests to pass, but the currently running ones seem stuck and the previous 6 tries all failed for different reasons unrelated to this PR. This one was the best: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8559/3874/summary.html. Only an unrelated llvm compiler warning and the PR to fix that was independently submitted already. (Core and DQM already approved except I put in a change to fix an unrelated unit test failure after that)

**Comment by slava77** [comment]
I echo David's request.
Changes are primarily technical.

**Comment by davidlange6** [comment]
All - i propose to merge this Saturday unless there are complaints in the meanwhile.

**Pull request merged**